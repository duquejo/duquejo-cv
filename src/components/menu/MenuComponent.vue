<template>
  <nav class="lg:flex lg:flex-col lg:items-center lg:justify-between hidden lg:my-4 relative">
    <LangSelectComponent id="menu-language-selector" />
    <div v-if="!isLoading && events" class="fixed w-96 h-auto top-0 bottom-0 my-4 rounded z-10 bg-gray-100 shadow-md flex flex-col justify-between gap-2 transition-all duration-500 ease-in-out" :class="menuClasses">
      <h4 class="mt-8 mb-2.5 text-h4 mx-4 text-center">{{ t('general.menu.events') }}</h4>
      <div class="overflow-y-auto">
          <RecentCardComponent v-for="event in events" :key="event.id" :event="event" />
          <p class="mx-4 mt-0 text-xs text-center">
            <a href="https://github.com/duquejo01" target="_blank" :aria-label="t('general.menu.events_repo_general_link_label')">{{ t('general.menu.events_repo_general_link') }}</a>
          </p>
      </div>
    </div>
    <i v-if="!isLoading" class="icon" @click="onLoadClick">
      <v-icon :class="eventIconClass" :name="eventIcon" scale="1.5" fill="black"/>
    </i>
    <i v-else class="icon cursor-default" :title="t('general.menu.events_loading_description')">
      <v-icon name="fa-spinner" scale="1" fill="gray" animation="spin"/>
    </i>
  </nav>
</template>

<script setup lang="ts">
import { computed, defineAsyncComponent } from 'vue';
import { useI18n } from 'vue-i18n';

import useQueryClient from '@/composables/useQueryClient';

const LangSelectComponent = defineAsyncComponent(() => import('@/components/language-select/LangSelectComponent.vue'));
const RecentCardComponent = defineAsyncComponent(() => import('@/components/recent-card/RecentCardComponent.vue'));

interface Props {
  isOpen?: boolean;
}

interface Emits {
  (e: 'toggle' ): void;
}

const { t } = useI18n();
const { events, isLoading, refetch } = useQueryClient();

const props = withDefaults(defineProps<Props>(), {
  isOpen: false,
});
const emit = defineEmits<Emits>();

const menuClasses = computed(() => ({
  'opacity-0 translate-x-full right-0': ! props.isOpen,
  'opacity-100 translate-x-0 right-12': props.isOpen,
}));

const eventIcon = computed(() => props.isOpen ? 'hi-solid-light-bulb' : 'hi-light-bulb' );
const eventIconClass = computed(() => ({
  'animate-bounce': ! props.isOpen,
  'animate-pulse': props.isOpen,
}));

const onLoadClick = () => {
  refetch();
  emit('toggle');
};

</script>