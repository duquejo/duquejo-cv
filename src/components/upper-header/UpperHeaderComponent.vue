
<template>
  <div class="sticky lg:relative flex justify-between w-full p-2 lg:hidden items-center">
    <span class="p-4" />
    <NameTagComponent classes="before:bg-gray-100"/>
    <v-icon :name="menuIcon" scale="1.5" @click="emit('menu-click')" />
  </div>
</template>

<script setup lang="ts">
import { computed } from 'vue';
import NameTagComponent from '@/components/name-tag/NameTagComponent.vue';

interface Props {
  isOpen: boolean;
}

interface Emits {
  (e: 'menu-click'): void;
}

const props = defineProps<Props>();
const emit = defineEmits<Emits>();

const menuIcon = computed(() => props.isOpen ? 'co-hamburger-menu' : 'pr-times' );
</script>