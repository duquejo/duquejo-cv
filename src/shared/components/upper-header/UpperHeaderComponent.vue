
<template>
  <div class="flex justify-between w-full p-2 lg:hidden items-center">
    <span class="p-4" />
    <span class="relative w-[max-content] font-mono before:absolute before:inset-0 before:animate-typewriter before:bg-gray-100 before:opacity-80 after:absolute after:inset-0 after:w-[0.125em] after:animate-caret after:bg-black tracking-widest">@duquejo</span>
    <v-icon :name="menuIcon" scale="1.5" />
  </div>
</template>

<script setup lang="ts">
import { computed } from 'vue';
interface Props {
  isOpen: boolean;
}
const props = defineProps<Props>();
const menuIcon = computed(() => props.isOpen ? 'co-hamburger-menu' : 'pr-times' );
</script>