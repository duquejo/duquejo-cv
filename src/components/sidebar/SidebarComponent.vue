<template>
  <aside id="sidebar" class="transition-all duration-500 absolute flex flex-col items-center justify-between lg:px-8 lg:py-10 lg:my-2 lg:mx-2 rounded-lg bg-white lg:h-auto z-10 w-full lg:w-auto lg:relative lg:opacity-100 lg:left-0 h-screen" :class="classes">
    <UpperHeaderComponent :isOpen="isOpen" @click="emit('toggle')"/>
    <div class="flex flex-col justify-center items-center w-full lg:p-0 px-8 py-10 text-center text-base">
      <picture class="mb-5">
        <img class="rounded-full w-32 h-32 sm:w-40 sm:h-40 object-cover grayscale hover:grayscale-0 transition-all duration-200" :src="smallAvatar" alt="José Duque" />
      </picture>
      <NameTagComponent classes="mb-1 before:bg-transparent text-gray-700 hidden lg:block" size="small"/>
      <span class="text-2xl font-semibold sm:text-xl">José Duque</span>
      <span class="text-gray-500">{{ t('general.role') }}</span>

      <small class="border-t border-gray-200 pt-px mt-5 mb-0 lg:hidden">{{ t('general.sidebar.network_links') }}</small>
      <div v-if="props.socialLinks?.length" class="flex flex-row justify-evenly lg:my-2 gap-x-3">
        <i class="icon" v-for="link in props.socialLinks" :key="link.name">
          <a :href="link.url" target="_blank">
            <v-icon :name="link.icon" scale="1.3" class="fill-gray-400 hover:fill-yellow-500 transition-colors" />
          </a>
        </i>
      </div>

      <small class="border-t border-gray-200 pt-px mt-3 mb-0 lg:hidden">{{ t('general.sidebar.theme_options') }}</small>
      <LangSelectComponent id="sidebar-language-selector" class="lg:hidden" />
    </div>
    <NavigationLinks @link-click="emit('toggle')" />
  </aside>
</template>

<script setup lang="ts">
import { computed } from 'vue';
import { useI18n } from 'vue-i18n';

import { SocialLinks } from '@/interfaces';

import UpperHeaderComponent from '@/components/upper-header/UpperHeaderComponent.vue';
import LangSelectComponent from '@/components/language-select/LangSelectComponent.vue';
import NameTagComponent from '@/components/name-tag/NameTagComponent.vue';
import NavigationLinks from '@/components/navigation-links/NavigationLinks.vue';

import smallAvatar from '@/assets/avatar-small.jpg';

interface Props {
  isOpen: boolean;
  socialLinks?: SocialLinks[];
}

interface Emit {
  (e: 'toggle'): void;
}

const { t } = useI18n();
const props = defineProps<Props>();
const emit = defineEmits<Emit>();

const classes = computed(() => ({
  'opacity-0 left-full': props.isOpen,
  'opacity-100 left-0': ! props.isOpen,
}));

</script>