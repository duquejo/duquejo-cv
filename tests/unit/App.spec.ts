import { beforeEach, describe, expect, it, vi } from 'vitest';
import { VueWrapper, shallowMount } from '@vue/test-utils';
import AppComponent from '../../src/App.vue';
import createVueI18n from '../mocks/mock-i18n';
import MenuComponent from '../../src/components/menu/MenuComponent.vue';
import SidebarComponent from '../../src/components/sidebar/SidebarComponent.vue';
import UpperHeaderComponent from '../../src/components/upper-header/UpperHeaderComponent.vue';


describe('+ App.vue tests', () => {

  const i18n = createVueI18n();

  let wrapper: VueWrapper;
  let typedWrapperVm: any;

  beforeEach(() => {
    vi.clearAllMocks();
    wrapper = shallowMount(AppComponent, {
      global: {
        plugins: [i18n],
        stubs: {
          'router-view': true
        }
      }
    });
    typedWrapperVm = (wrapper.vm as any);
  });

  it('should match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.findComponent(MenuComponent)).toBeTruthy();
    expect(wrapper.findComponent(SidebarComponent)).toBeTruthy();
    expect(wrapper.findComponent(UpperHeaderComponent)).toBeTruthy();
  });

  it('should toggle isOpen when onToggleMenu is called', () => {
    typedWrapperVm = (wrapper.vm as any);
    typedWrapperVm.onToggleMenu()
    expect(typedWrapperVm.isOpen).toBe(false);
    typedWrapperVm.onToggleMenu();
    expect(typedWrapperVm.isOpen).toBe(true);
  });

  it('should toggle isOpenMenu when onToggleSidebarMenu is called', () => {
    typedWrapperVm.onToggleSidebarMenu();
    expect(typedWrapperVm.isOpenMenu).toBe(true);
    typedWrapperVm.onToggleSidebarMenu();
    expect(typedWrapperVm.isOpenMenu).toBe(false);
  });

  it('should set isOpenMenu to false when onOutsideClick is called and isOpenMenu is true', () => {
    typedWrapperVm.isOpenMenu = true;
    typedWrapperVm.onOutsideClick();
    expect(typedWrapperVm.isOpenMenu).toBe(false);
  });

  it('should not change isOpenMenu when onOutsideClick is called and isOpenMenu is false', () => {
    typedWrapperVm.isOpenMenu = false;
    typedWrapperVm.onOutsideClick();
    expect(typedWrapperVm.isOpenMenu).toBe(false);
  });

  it('should compute mainClasses correctly', () => {
    typedWrapperVm.isOpenMenu = true;
    expect(typedWrapperVm.mainClasses).toEqual({
      'blur-sm': true,
    });
    typedWrapperVm.isOpenMenu = false;
    expect(typedWrapperVm.mainClasses).toEqual({
      'blur-sm': false,
    });
  });
});