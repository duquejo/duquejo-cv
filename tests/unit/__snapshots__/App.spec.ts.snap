// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`+ App.vue tests > should match the snapshot 1`] = `
"<sidebar-component-stub data-v-7a7a37b1="" sociallinks="[object Object],[object Object]" isopen="true"></sidebar-component-stub>
<upper-header-component-stub data-v-7a7a37b1="" isopen="true"></upper-header-component-stub>
<main data-v-7a7a37b1="" class="overflow-y-scroll bg-white transition-all duration-1000 relative">
  <router-view-stub data-v-7a7a37b1=""></router-view-stub>
</main>
<navigation-links-stub data-v-7a7a37b1="" layout="fixed" size="small"></navigation-links-stub>
<menu-component-stub data-v-7a7a37b1="" isopen="false"></menu-component-stub>"
`;
